Sources/MongoSwift/APM.swift

```diff
@@ -0,0 +1,228 @@
+    let operationId: Int64?
+
+    /// The connection id for the command.
+    let connectionId: ConnectionId
+
+    /// An internal initializer for creating a CommandStartedEvent from a mongoc_apm_command_started_t
+    internal init(_ event: OpaquePointer) {
+        let commandData = UnsafeMutablePointer(mutating: mongoc_apm_command_started_get_command(event)!)
+        self.command = Document(fromPointer: commandData)
+        self.databaseName = String(cString: mongoc_apm_command_started_get_database_name(event))
+        self.commandName = String(cString: mongoc_apm_command_started_get_command_name(event))
+        self.requestId = mongoc_apm_command_started_get_request_id(event)
+        self.operationId = mongoc_apm_command_started_get_operation_id(event)
+        self.connectionId = ConnectionId(mongoc_apm_command_started_get_host(event))
+    }
+}
+
+/// An event published when a command succeeds. The event is stored under the key `event`
+/// in the `userInfo` property of `Notification`s posted under the name .commandSucceeded.
+public struct CommandSucceededEvent {
+    /// The execution time of the event, in microseconds.
+    let duration: Int64
+
+    /// The command reply.
+    let reply: Document
+
+    /// The command name.
+    let commandName: String
+
+    /// The driver generated request id.
+    let requestId: Int64
+
+    /// The driver generated operation id. This is used to link events together such
+    /// as bulk write operations.
+    let operationId: Int64?
+
+    /// The connection id for the command.
+    let connectionId: ConnectionId
+
+    /// An internal initializer for creating a CommandSucceededEvent from a mongoc_apm_command_succeeded_t
+    internal init(_ event: OpaquePointer) {
+        self.duration = mongoc_apm_command_succeeded_get_duration(event)
+        let replyData = UnsafeMutablePointer(mutating: mongoc_apm_command_succeeded_get_reply(event)!)
+        self.reply = Document(fromPointer: replyData)
+        self.commandName = String(cString: mongoc_apm_command_succeeded_get_command_name(event))
+        self.requestId = mongoc_apm_command_succeeded_get_request_id(event)
+        self.operationId = mongoc_apm_command_succeeded_get_operation_id(event)
+        self.connectionId = ConnectionId(mongoc_apm_command_succeeded_get_host(event))
+    }
+}
+
+/// An event published when a command fails. The event is stored under the key `event`
+/// in the `userInfo` property of `Notification`s posted under the name .commandFailed.
+public struct CommandFailedEvent {
+
+    /// The execution time of the event, in microseconds.
+    let duration: Int64
+
+    /// The command name.
+    let commandName: String
+
+    /// The failure, represented as a MongoError.
+    let failure: MongoError
+
+    /// The client generated request id.
+    let requestId: Int64
+
+    /// The driver generated operation id. This is used to link events together such
+    /// as bulk write operations.
+    let operationId: Int64?
```

Comment:
is this really optional?

Reply:
ditto for `CommandSucceededEvent`

Reply:
I took the optionality from the defined interfaces in the spec - it sounded like it would only be set for certain operation types.  but it appears that `libmongoc` always returns a value so I'll make it non-optional.